include/zephyr/virtio/virtio.h

```diff
@@ -0,0 +1,182 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Virtio Interface
+ * @defgroup virtio_interface Virtio Interface
+ * @{
+ */
+
+
+/**
+ * Callback used during virtqueue enumeration
+ *
+ * @param queue_idx index of currently inspected queue
+ * @param max_queue_size maximum permitted size of currently inspected queue
+ * @param opaque pointer to user provided data
+ * @return the size of currently inspected virtqueue we want to set
+ */
+typedef uint16_t (*virtio_enumerate_queues)(
+	uint16_t queue_idx, uint16_t max_queue_size, void *opaque
+);
+
+/**
+ * @brief Virtio api structure
+ */
+__subsystem struct virtio_driver_api {
+	struct virtq *(*get_virtqueue)(const struct device *dev, uint16_t queue_idx);
+	void (*notify_virtqueue)(const struct device *dev, uint16_t queue_idx);
+	void *(*get_device_specific_config)(const struct device *dev);
+	bool (*read_device_feature_bit)(const struct device *dev, int bit);
+	bool (*read_driver_feature_bit)(const struct device *dev, int bit);
+	int (*write_driver_feature_bit)(const struct device *dev, int bit, bool value);
+	int (*commit_feature_bits)(const struct device *dev);
+	int (*init_virtqueues)(const struct device *dev, virtio_enumerate_queues cb, void *opaque);
+	void (*finalize_init)(const struct device *dev);
+};
+
+/**
+ * Returns virtqueue at given idx
+ *
+ * @param dev virtio device it operates on
+ * @param queue_idx index of virtqueue to get
+ * @return pointer to virtqueue or NULL if not present
+ */
+static inline struct virtq *virtio_get_virtqueue(const struct device *dev, uint16_t queue_idx)
+{
+	const struct virtio_driver_api *api = dev->api;
+
+	return api->get_virtqueue(dev, queue_idx);
+}
+
+/**
+ * Notifies virtqueue
+ *
+ * @param dev virtio device it operates on
+ * @param queue_idx virtqueue to be notified
+ */
+static inline void virtio_notify_virtqueue(const struct device *dev, uint16_t queue_idx)
+{
+	const struct virtio_driver_api *api = dev->api;
+
+	api->notify_virtqueue(dev, queue_idx);
+}
+
+/**
+ * Returns device specific config
+ *
+ * @param dev virtio device it operates on
+ * @return pointer to the device specific config or NULL if its not present
+ */
+static inline void *virtio_get_device_specific_config(const struct device *dev)
+{
+	const struct virtio_driver_api *api = dev->api;
+
+	return api->get_device_specific_config(dev);
+}
+
+/**
+ * Returns feature bit offered by virtio device
+ *
+ * @param dev virtio device it operates on
+ * @param bit selected bit
+ * @return value of the offered feature bit
+ */
+static inline bool virtio_read_device_feature_bit(const struct device *dev, int bit)
+{
+	const struct virtio_driver_api *api = dev->api;
+
+	return api->read_device_feature_bit(dev, bit);
+}
+
+/**
+ * Returns feature bit selected by driver
+ *
+ * @param dev virtio device it operates on
+ * @param bit selected bit
+ * @return value of the selected bit
+ */
+static inline bool virtio_read_driver_feature_bit(const struct device *dev, int bit)
```

Comment:
As feedback from the result of implementing virtio-mmio in #89460.

The driver features register is

https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.pdf#subsection.4.2.2

```
DriverFeatures 0x020 W
```

and cannot be read. If you are going to abstract virtio-mmio, it may be better not to have it.